### PR TITLE
Fix issue when SDL shut down due to SIGSEGV

### PR DESCRIPTION
**Pull request by okozlovlux**
- replaced iterator by SharedPointer in CommandRequestImpl to
    access application

@LuxoftAKutsan  @LevchenkoS @indlu @dcherniev @LitvinenkoIra please, review.

### CODE REVIEW

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 1]
```diff
@@ -493,60 +493,58 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
-          if (true == iter->second.asBool()) {
-            LOG4CXX_DEBUG(logger_, "Request's param: " << iter->first);
-            params.insert(iter->first);
-          }
-        }
-      }
+  const ApplicationConstSharedPtr app =
+      application_manager_.application(connection_key());
+  if (!app) {
+    return false;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux please log Error

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 2]
```diff
@@ -493,60 +493,58 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
 
-      // If no parameters specified in policy table, no restriction will be
-      // applied for parameters
-      if (parameters_permissions_.allowed_params.empty() &&
-          parameters_permissions_.disallowed_params.empty() &&
-          parameters_permissions_.undefined_params.empty()) {
-        return true;
-      }
+  const smart_objects::SmartObject& s_map = (*message_)[strings::msg_params];
+  if (smart_objects::SmartType_Map == s_map.getType()) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux check is redundant. Our Smart object specification garantyy that it should be Map. You can DCHek it

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 3]
```diff
@@ -493,60 +493,58 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
+    smart_objects::SmartMap::iterator iter = s_map.map_begin();
+    smart_objects::SmartMap::iterator iter_end = s_map.map_end();
 
-      RemoveDisallowedParameters();
+    for (; iter != iter_end; ++iter) {
+      if (true == iter->second.asBool()) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux  if (iter->second.asBool()) {

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 4]
```diff
@@ -493,60 +493,58 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
-      // applied for parameters
-      if (parameters_permissions_.allowed_params.empty() &&
-          parameters_permissions_.disallowed_params.empty() &&
-          parameters_permissions_.undefined_params.empty()) {
-        return true;
-      }
+  const smart_objects::SmartObject& s_map = (*message_)[strings::msg_params];
+  if (smart_objects::SmartType_Map == s_map.getType()) {
+    smart_objects::SmartMap::iterator iter = s_map.map_begin();
+    smart_objects::SmartMap::iterator iter_end = s_map.map_end();
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux can be const

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 5]
```diff
@@ -493,60 +493,58 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
+    application_manager_.SendMessageToMobile(response);
+    return false;
+  }
+
+  // If no parameters specified in policy table, no restriction will be
+  // applied for parameters
+  if (parameters_permissions_.allowed_params.empty() &&
+      parameters_permissions_.disallowed_params.empty() &&
+      parameters_permissions_.undefined_params.empty()) {
+    return true;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux what does this check changes in behavior?

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 1 above)
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan  fixed 937b032

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 2 above)
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan  fixed 937b032

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 4 above)
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan  fixed 937b032

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 3 above)
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan  fixed 937b032

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 5 above)
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @LuxoftAKutsan , didn't add this functionality.

Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux but it was in another context. Please clarify

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 6]
```diff
@@ -493,60 +493,61 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
+  smart_objects::SmartMap::const_iterator iter_end = s_map.map_end();
+
+  for (; iter != iter_end; ++iter) {
+    if (iter->second.asBool()) {
+      LOG4CXX_DEBUG(logger_, "Request's param: " << iter->first);
+      params.insert(iter->first);
     }
   }
+
+  mobile_apis::Result::eType check_result =
```
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @okozlovlux, may be const?

**Comment by okozlovlux** [comment]
Rebuild required

**Comment by okozlovlux** [comment]
Rebuild required

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 7]
```diff
@@ -493,60 +493,61 @@ bool CommandRequestImpl::CheckAllowedParameters() {
...
+      application_manager_.CheckPolicyPermissions(
+          app->policy_app_id(),
+          app->hmi_level(),
+          static_cast<mobile_api::FunctionID::eType>(function_id()),
+          params,
+          &parameters_permissions_);
+
+
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/command_request_impl.cc]: The `check_style.sh` script is pointing out that there should only be one newline here, not two.

**Comment by LuxoftAKutsan** [comment]
@jacobkeeler please attach more ATF artifacts for analyses.

**Comment by LuxoftAKutsan** [comment]
@jacobkeeler I checked on Ubuntu 16.04.1 LTS
It is SUCCESS:
```
rm TestingReports -rf ;  ./start.sh --sdl-core=/tmp/okozlov test_scripts/Polices/Validation_of_PolicyTables/Check_count_of_removals_for_bad_behavior_REQUEST_WHILE_IN_NONE_HMI_LEVEL.lua 
==============================
Start 'test_scripts/Polices/Validation_of_PolicyTables/Check_count_of_removals_for_bad_behavior_REQUEST_WHILE_IN_NONE_HMI_LEVEL.lua'
==============================
 WARNING: Parameter pathToSDLInterfaces is not specified, default APIs are used 
SDL pid 11349
[11:58:18,101] RunSDL                                                                                [SUCCESS] (2 ms)
[11:58:19,112] InitHMI                                                                               [SUCCESS] (681 ms)
[11:58:19,823] InitHMI_onReady                                                                       [SUCCESS] (822 ms)
[11:58:20,656] ConnectMobile                                                                         [SUCCESS] (7 ms)
[11:58:20,665] StartSession                                                                          [SUCCESS] (2034 ms)
--------------------------------------------------------------------------------


Preconditions
--------------------------------------------------------------------------------
[11:58:22,703] Suite_1                                                                               [SUCCESS] (388 ms)
[11:58:23,092] Precondition_StopSDL                                                                  [SUCCESS] (1 ms)
[11:58:24,126] ChangeIniFile                                                                         [SUCCESS] (2 ms)
SDL pid 11614
[11:58:24,136] Precondition_StartSDL                                                                 [SUCCESS] (2 ms)
[11:58:25,152] Precondition_initHMI                                                                  [SUCCESS] (1287 ms)
[11:58:26,628] Precondition_initHMI_onReady                                                          [SUCCESS] (1060 ms)
[11:58:27,720] Precondition_ConnectMobile                                                            [SUCCESS] (8 ms)
[11:58:27,776] Precondition_StartSession                                                             [SUCCESS] (84 ms)
[11:58:27,872] RegisterApp                                                                           [SUCCESS] (553 ms)
[11:58:28,588] Send_TOO_MANY_REQUESTS_WHILE_IN_NONE_HMI_LEVEL                                        [SUCCESS] (174 ms)
[11:58:29,133] Wait                                                                                  [SUCCESS] (5 ms)
[11:58:32,159] Check_TOO_MANY_REQUESTS_in_DB                                                         [SUCCESS] (4 ms)
[11:58:33,173] Postcondition_Stop_SDL                                                                [SUCCESS] (1 ms)
[11:58:34,204] RestoreIniFile                                                                        [SUCCESS] (2 ms)
Total executing time is 16s 179ms (summary 16179ms)
==============================
Finish 'test_scripts/Polices/Validation_of_PolicyTables/Check_count_of_removals_for_bad_behavior_REQUEST_WHILE_IN_NONE_HMI_LEVEL.lua'
➜  test_run lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.1 LTS
Release:        16.04
Codename:       xenial
[TestingReports.tar.gz](https://github.com/smartdevicelink/sdl_core/files/992417/TestingReports.tar.gz)

```
****
ATF Commit : fe002fb 
SDL Commit : 937b0323ce2fed5fbd5cfb86c4a2b4b966ba45e4
ATF scripts commit : 1435aa8

File: src/components/application_manager/src/commands/command_request_impl.cc (hunk 7 above)
Comment (okozlovlux) [inline src/components/application_manager/src/commands/command_request_impl.cc]: @jacobkeeler , fixed coding style in 2520ea0

**Comment by okozlovlux** [comment]
Rebuild required

**Comment by jacobkeeler** [comment]
@LuxoftAKutsan Here is the full test log:
```
./start.sh test_scripts/Polices/Validation_of_PolicyTables/Check_count_of_removals_for_bad_behavior_REQUEST_WHILE_IN_NONE_HMI_LEVEL.lua 
==============================
Start 'test_scripts/Polices/Validation_of_PolicyTables/Check_count_of_removals_for_bad_behavior_REQUEST_WHILE_IN_NONE_HMI_LEVEL.lua'
==============================
qt.network.ssl: QSslSocket: cannot resolve SSLv2_client_method
qt.network.ssl: QSslSocket: cannot resolve SSLv2_server_method
SDL pid 31132
[09:07:19,984] RunSDL                                                                                [SUCCESS] (390 ms)
accept_buf: 9wSSP8SIkCON/d/+vFVdxAbUIpQ=
[09:07:20,391] InitHMI                                                                               [SUCCESS] (150 ms)
[09:07:20,542] InitHMI_onReady                                                                       [SUCCESS] (110 ms)
[09:07:20,660] ConnectMobile                                                                         [SUCCESS] (7 ms)
[09:07:20,669] StartSession                                                                          [SUCCESS] (2030 ms)
--------------------------------------------------------------------------------


Preconditions
--------------------------------------------------------------------------------
[09:07:22,701] Suite_1                                                                               [SUCCESS] (140 ms)
[09:07:22,842] Precondition_StopSDL                                                                  [SUCCESS] (2 ms)
[09:07:23,874] ChangeIniFile                                                                         [SUCCESS] (2 ms)
SDL pid 31390
[09:07:23,881] Precondition_StartSDL                                                                 [SUCCESS] (3 ms)
 SDL didn't send anything for 8000 msecs. Closing session # 1
[09:07:23,904] Precondition_initHMI                                                                  [FAIL] (18739 ms)
 Connected websocket: Timeout: Connected websocket: Timeout expired
 EndService ACK: Timeout: EndService ACK: Timeout expired
[09:07:42,645] Precondition_initHMI_onReady                                                          [FAIL] (10765 ms)
 VehicleInfo.GetVehicleData: Timeout: VehicleInfo.GetVehicleData: Timeout expired
 TTS.GetSupportedLanguages: Timeout: TTS.GetSupportedLanguages: Timeout expired
 TTS.IsReady: Timeout: TTS.IsReady: Timeout expired
 VehicleInfo.IsReady: Timeout: VehicleInfo.IsReady: Timeout expired
 UI.GetCapabilities: Timeout: UI.GetCapabilities: Timeout expired
 VR.GetLanguage: Timeout: VR.GetLanguage: Timeout expired
 UI.IsReady: Timeout: UI.IsReady: Timeout expired
 UI.GetSupportedLanguages: Timeout: UI.GetSupportedLanguages: Timeout expired
 VehicleInfo.GetVehicleType: Timeout: VehicleInfo.GetVehicleType: Timeout expired
 VR.IsReady: Timeout: VR.IsReady: Timeout expired
 Navigation.IsReady: Timeout: Navigation.IsReady: Timeout expired
 TTS.GetLanguage: Timeout: TTS.GetLanguage: Timeout expired
 TTS.GetCapabilities: Timeout: TTS.GetCapabilities: Timeout expired
 BasicCommunication.MixingAudioSupported: Timeout: BasicCommunication.MixingAudioSupported: Timeout expired
 Buttons.GetCapabilities: Timeout: Buttons.GetCapabilities: Timeout expired
 UI.GetLanguage: Timeout: UI.GetLanguage: Timeout expired
 VR.GetCapabilities: Timeout: VR.GetCapabilities: Timeout expired
 VR.GetSupportedLanguages: Timeout: VR.GetSupportedLanguages: Timeout expired
[09:07:53,446] Precondition_ConnectMobile                                                            [SUCCESS] (5 ms)
[09:07:53,455] Precondition_StartSession                                                             [SUCCESS] (384 ms)
 SDL didn't send anything for 8000 msecs. Closing session # 1
[09:07:53,843] RegisterApp                                                                           [FAIL] (10798 ms)
 OnPermissionsChange notification: Timeout: OnPermissionsChange notification: Timeout expired
 response to 2: Timeout: response to 2: Timeout expired
 HMI notification BasicCommunication.OnAppRegistered: Timeout: HMI notification BasicCommunication.OnAppRegistered: Timeout expired
[09:08:04,647] Send_TOO_MANY_REQUESTS_WHILE_IN_NONE_HMI_LEVEL                                        [FAIL] (10782 ms)
 OnAppInterfaceUnregistered notification: Timeout: OnAppInterfaceUnregistered notification: Timeout expired
[09:08:15,445] Wait                                                                                  [SUCCESS] (5 ms)
[09:08:18,454] Check_TOO_MANY_REQUESTS_in_DB                                                         [FAIL] (1020 ms)
 AutoFail: 1: DB doesn't include expected value
[09:08:19,476] Postcondition_Stop_SDL                                                                [SUCCESS] (1 ms)
[09:08:20,501] RestoreIniFile                                                                        [SUCCESS] (1 ms)
Total executing time is 1min 571ms (summary 60571ms)
==============================
Finish 'test_scripts/Polices/Validation_of_PolicyTables/Check_count_of_removals_for_bad_behavior_REQUEST_WHILE_IN_NONE_HMI_LEVEL.lua'
==============================
```

[TestingReports.zip](https://github.com/smartdevicelink/sdl_core/files/993088/TestingReports.zip)

**Comment by okozlovlux** [comment]
@jacobkeeler , please give the ATF commit and ATF scripts commit that you use.

**Comment by jacobkeeler** [comment]
@okozlovlux Develop branch for both:
ATF commit: https://github.com/smartdevicelink/sdl_atf/commit/fe002fb0e4e79cf9487f9c8bda6af7d61198a26d
ATF Test Scripts commit: https://github.com/smartdevicelink/sdl_atf_test_scripts/commit/599e1eb1738285bc3e2631d16e1fd77a6b191fe0

**Pull request merged**